pyriemann/utils/rotate.py

```diff
@@ -0,0 +1,290 @@
+import numpy as np
+from ..utils.distance import distance_riemann, distance_euclid
+
+
+def _project(X, U):
+
+    """
+    Project U on the tangent space of orthogonal matrices with base point X
+    """
+
+    return X @ ((X.T @ U) - (X.T @ U).T) / 2
+
+
+def _retract(X, v):
+
+    """
+    Retraction taking tangent vector v at base point X back to manifold of
+    orthogonal matrices
+    """
+
+    Q, R = np.linalg.qr(X + v)
+    for i, ri in enumerate(np.diag(R)):
+        if ri < 0:
+            Q[:, i] = -1 * Q[:, i]
+    return Q
+
+
+def _check_dimensions(M):
+    if isinstance(M, list):
+        M = np.stack(M)
+    if isinstance(M, np.ndarray):
+        if M.ndim == 2:
+            M = M[None, :, :]
+    return M
+
+
+def _loss_rie(Q, X, Y, weights=None):
+
+    """
+    Loss function for estimating the rotation matrix in RPA using the
+    riemannian distance between the class means.
+    """
+
+    def loss_i_rie(Q, Xi, Yi):
+        return distance_riemann(Xi, Q @ Yi @ Q.T)**2
+
+    X = _check_dimensions(X)
+    Y = _check_dimensions(Y)
+
+    if len(X) != len(Y):
+        raise ValueError("The number of classes in each domain don't match")
+
+    if weights is None:
+        weights = np.ones(len(X)) / len(X)
+
+    L = 0
+    for i in range(len(X)):
+        L = L + weights[i] * loss_i_rie(Q, X[i], Y[i])
+
+    return L
+
+
+def _grad_rie(Q, X, Y, weights=None):
+
+    """
+    Gradient of loss function using Riemannian distances between class means.
+    """
+
+    def _grad_i_rie(Q, Xi, Yi):
+        M = np.linalg.inv(Xi) @ Q @ Yi @ Q.T
+        w, v = np.linalg.eig(M)
+        logM = v @ np.diag(np.log(w)) @ np.linalg.inv(v)
```

Comment:
It is a matrix with positive eigenvalues but not symmetric. Using `utils.base.logm` throws an error because it is implemented using `np.linalg.eigh` which gives rubbish when you input a non-hermitian matrix.